crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+            if (calc_priv == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            calc_pub = EC_POINT_new(curve);
+            if (calc_pub == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if (EC_POINT_mul(curve, calc_pub, calc_priv, NULL, NULL,
+                             NULL) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if ((calc_pubuf_len = EC_POINT_point2oct(curve, calc_pub, form,
+                                                     calc_pubuf, calc_pubuf_len,
+                                                     NULL)) != pubsize) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if (OSSL_PARAM_BLD_push_octet_string(param_bld, "pub", calc_pubuf,
+                                                 calc_pubuf_len) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+#else
+            /* can't do that if no EC support compiled in:-( */
+            OSSL_HPKE_err;
+            goto err;
+#endif
+        }
+        if (strlen(keytype) == 2 && !strcmp(keytype, "EC")) {
+            priv = BN_bin2bn(prbuf, prbuf_len, NULL);
+            if (priv == NULL) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+            if (OSSL_PARAM_BLD_push_BN(param_bld, "priv", priv) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+        } else {
+            if (OSSL_PARAM_BLD_push_octet_string(param_bld, "priv", prbuf,
+                                                 prbuf_len) != 1) {
+                OSSL_HPKE_err;
+                goto err;
+            }
+        }
+        params = OSSL_PARAM_BLD_to_param(param_bld);
+        if (params == NULL) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        ctx = EVP_PKEY_CTX_new_from_name(libctx, keytype, NULL);
+        if (ctx == NULL) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        if (EVP_PKEY_fromdata_init(ctx) <= 0) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        if (EVP_PKEY_fromdata(ctx, &lpriv, EVP_PKEY_KEYPAIR, params) <= 0) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+    }
+    if (lpriv == NULL) {
+        /* check PEM decode - that might work :-) */
+        BIO *bfp = BIO_new(BIO_s_mem());
+
+        if (bfp == NULL) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        BIO_write(bfp, prbuf, prbuf_len);
+        if (!PEM_read_bio_PrivateKey(bfp, &lpriv, NULL, NULL)) {
+            BIO_free_all(bfp);
+            bfp = NULL;
+            OSSL_HPKE_err;
+            goto err;
+        }
+        if (bfp != NULL) {
+            BIO_free_all(bfp);
+            bfp = NULL;
+        }
+        if (lpriv == NULL) {
+            /* if not done, prepend/append PEM header/footer and try again */
```

Comment:
Why is this necessary?

Reply:
This is basically the same issue noted above wrt OSSL_HPKE_prbuf2evp flexibility.